### PR TITLE
Update TAD guidelines based on notes from Engineering Meeting discussion

### PR DESCRIPTION
**Pull request by samanthaandrews**
I facilitated two TAD-focused discussions this year during the bi-monthly Engineering Meetings. The changes are a result of those discussions, as noted in the [notes document](https://docs.google.com/document/d/1BHSAW8sNNebiuqNQoIPFgj-HJI4LdQa5CmycrJ6LBYo/edit#heading=h.b5fuy6svo0bi) from Sept 6 and Sept 20, 2023.

### CODE REVIEW

**Review by hokei — APPROVED** [review]
This looks good to me reflects a lot of my own thoughts on TADs.  One thing that would be really helpful is to link to certain examples of good TADs as the TAD drive has a lot of documents, and some approved TADs that don't conform to all these principles.  I feel like I might be unusual in that I came into this document already clear on what I'm looking for out of a TAD, and everything here matches what I'd expect to see, but I'd love to get some feedback from people who feel vague about TAD requirements on what else might need clarifying.

**Comment by samanthaandrews** [comment]
Thanks, @hokei 

> One thing that would be really helpful is to link to certain examples of good TADs as the TAD drive has a lot of documents, and some approved TADs that don't conform to all these principles.

I added a few examples of strong TADs and ones with opportunities for improvement, based on my experience with eReading TADs. Would you and @mwbenowitz also send me some links to TADs with a sentence calling out what conforms to and doesn't conform to the spec? It's difficult for me to tell from the docs in the drive which are _final_ and which are still drafts or copies 😅 I'll also ask the rest of the ELT folks to contribute!

**Comment by samanthaandrews** [comment]
@nonword @nypl-wluisi @emu47 @EdwinGuzman @russellcullen 

Please send me some examples of strong TADs and some examples of TADs that don't quite hit the mark so that I can add them to the list of examples!

File: standards/technical-approach.md
Comment (EdwinGuzman) [inline standards/technical-approach.md:81]: This is great. I've also been thinking of this as an "Addendum" where we write updates to the TAD based on new findings.

File: standards/technical-approach.md
Comment (russellcullen) [inline standards/technical-approach.md:59]: Would another branch in the same repo also be acceptable here?

File: standards/technical-approach.md
Comment (russellcullen) [inline standards/technical-approach.md:70]: nit: Should we use Garvita's title here in case the exact person we send TADs to will change?

**Review by russellcullen — APPROVED** [review]
Love it! Thanks for updating

**Comment by gioneill** [comment]
LGTM! 🤣🤫On Jan 5, 2024, at 8:22 AM, Edwin Guzman ***@***.***> wrote:﻿
@EdwinGuzman commented on this pull request.



In standards/technical-approach.md:
> @@ -43,8 +78,24 @@ Step 5 above includes the following steps to be followed by the Tech Lead/Archit
 2. If revisions are requested/necessary the document's author should make the necessary changes and record the process in a changelog to be added to the end of the document
 3. The above steps should be repeated until the document is approved
 
+### Changes After Approval

This is great. I've also been thinking of this as an "Addendum" where we write updates to the TAD based on new findings.

**Review by EdwinGuzman — APPROVED** [review]

File: standards/technical-approach.md
Comment (samanthaandrews) [reply standards/technical-approach.md:70]: good idea!

File: standards/technical-approach.md
Comment (samanthaandrews) [reply standards/technical-approach.md:59]: Yes i think so. Updated!

File: standards/technical-approach.md
Comment (Toxiapo) [inline standards/technical-approach.md:83]: Q: Is there specific format we should follow for the changelog section? How do we communicate this to ELT/leads about the change. (i.e. Do we need another approval for pivoting to a different approach?)

Comment (samanthaandrews) [reply standards/technical-approach.md:83]: I did update the TAD template with a Changelog section at the bottom, which right now only includes a date heading.
Screenshot 2024-01-08 at 6 22 02 PM

Updating the TAD's changelog should include communicating with your Tech Lead or Architect, and from there Crystal or Mike should probably be updated. I'm imagining this happening throughout the regular 1:1 meeting, so I'll add this process here. Good idea.

Comment (Toxiapo) [reply standards/technical-approach.md:83]: Thanks for the clarifications!

File: standards/technical-approach.md
Comment (ettore) [inline standards/technical-approach.md:9]: one of the main challenges I encountered when I wrote TADs is figuring out the audience I was writing to. 

My initial assumption was that I was writing for my peers, so in my case iOS developers. So I would not add in my TAD explanations about certain iOS technologies or why other things are ok or not ok to use, because they are obvious in iOS development. But then eventually I had to rewrite entire paragraphs to make them understandable to engineers not familiar with the iOS stack. I have no problem in doing that btw: my point here is that saying "technical audience" is kind of vague and may include people with completely different technical training.

I know that "technical audience" includes engineers that will implement the feature, so in my example, iOS engineers. But that's not the only audience. What can we assume for this other audience?

**Review by ettore — COMMENTED** [review]
I really like this improved document and it clarifies some of my doubts. I especially love the examples section! I wish I had good examples when I wrote my TADs.

File: standards/technical-approach.md
Comment (oliviawongnyc) [inline standards/technical-approach.md:6]: I think the first two paragraphs could be a little more concise and direct about what the document is. What do you think? Maybe something along the lines of:

"Technical Approach Documents (TADs) allow engineers to explore multiple possible solutions to a technical problem before development begins, uncovering potential issues and ultimately finding a recommended approach through the process."

File: standards/technical-approach.md
Comment (oliviawongnyc) [reply standards/technical-approach.md:66]: A related thought – should we recommend a way of pointing TADs?

File: standards/technical-approach.md
Comment (oliviawongnyc) [inline standards/technical-approach.md:70]: Nit: This comma should be moved to after "in Step 5" or just remove "in Step 5" altogether.

Comment (oliviawongnyc) [reply standards/technical-approach.md:70]: Nit: move this line, "With final approval the associated JIRA ticket is moved to `Done`" to the next step.

"7. With final approval the associated JIRA ticket is moved to `Done`. Specific tickets MUST be created to execute the work as described in the TAD. "

File: standards/technical-approach.md
Comment (oliviawongnyc) [inline standards/technical-approach.md:83]: Nit: typo in "existing", shorten last sentence for clarity:

"You should also tell your Tech Lead/Architect/Manager about these updates so they can be communicated up the chain if necessary."

File: standards/technical-approach.md
Comment (oliviawongnyc) [reply standards/technical-approach.md:9]: I agree it would be helpful to have more clarity on this. Is ELT different from DLT? Maybe ELT should be defined.

**Review by oliviawongnyc — COMMENTED** [review]
Looking really good! One last thought – what do you think about adding a section about the template? We talk a lot about the peripheral pieces of the TAD - timing, approval, authorship, etc, but not much about its actual content. I think the examples help, but maybe could add a paragraph on the pieces of the template. For example, writers should start by stating the problem in a couple of sentences, that an architectural summary is expected and what that means, one should consider backend, frontend, infrastructure and security concerns...

File: standards/technical-approach.md
Comment (samanthaandrews) [reply standards/technical-approach.md:43]: this is great, thank you!

File: standards/technical-approach.md
Comment (samanthaandrews) [reply standards/technical-approach.md:6]: Agreed, this is better. Thank you.

File: standards/technical-approach.md
Comment (samanthaandrews) [inline standards/technical-approach.md:11]: @ettore @oliviawongnyc let me know what you think of this in clarifying ELT! basically ELT is (currently) Garvita, Crystal, Mike, and Anthony. Anthony doesn't review these TADs, but Crystal, Mike, and Garvita do.

File: standards/technical-approach.md
Comment (samanthaandrews) [inline standards/technical-approach.md:64]: @oliviawongnyc Added this section about the template, which I do think is helpful. Lmk what you think.

File: standards/technical-approach.md
Comment (samanthaandrews) [inline standards/technical-approach.md:84]: @oliviawongnyc added some detail here about pointing TADs. The recent PJR TADs that I have seen are all pointed at 5 points, which seems standard, but I don't think 8 is unreasonable.

File: standards/technical-approach.md
Comment (samanthaandrews) [inline standards/technical-approach.md:85]: @ettore how does this look?

File: standards/technical-approach.md
Comment (samanthaandrews) [inline standards/technical-approach.md:3]: @oliviawongnyc shortened this up as you suggested

File: standards/technical-approach.md
Comment (ettore) [reply standards/technical-approach.md:11]: This sounds good to me, so if I understand correctly we should add more context (i.e. assume less knowledge of the stack) in those 2 sections (Problem def/Goal and Architectural Summary) but for the rest our assumption should be the people who routinely review our PRs.  Did I understand well?

File: standards/technical-approach.md
Comment (ettore) [reply standards/technical-approach.md:85]: looks great!

**Review by ettore — APPROVED** [review]
thank you for all the fixes @samanthaandrews ! This looks really good to me.

File: standards/technical-approach.md
Comment (Apophenia) [inline standards/technical-approach.md:88]: Seems like some formatting/line breaks got lost with this update

**Comment by Apophenia** [comment]
I think this looks really great, and virtually all of the questions/suggestions I had have been well addressed. Barring what I think is a line break typo in that one section, this looks good to me!

File: standards/technical-approach.md
Comment (samanthaandrews) [reply standards/technical-approach.md:88]: woah you're right good catch

File: standards/technical-approach.md
Comment (oliviawongnyc) [reply standards/technical-approach.md:64]: Excellent! TY!

**Review by Apophenia — APPROVED** [review]
Amazing!

**Review by Toxiapo — APPROVED** [review]

File: standards/technical-approach.md
Comment (nonword) [inline standards/technical-approach.md:84]: I think team points allotment is bound to differ and some TADs are pretty research and prototype intensive, so 5-8 seems low to me based on what we've done in LSP. I don't think there are many LSP TADs that we would call 8pts after the fact - closer to 13 when you include discussions and edits.

File: standards/technical-approach.md
Comment (nonword) [inline standards/technical-approach.md:27]: I'll often include appendices with supplemental data, code samples, and other asides that aren't essential reading (but linked from the body of the TAD to support certain ideas). I sometimes prefer this to placing that stuff in outside, linked documents (although I also sometimes link to outside, supporting spreadsheets and the like). I'm hoping folks will discount appendices content when assessing page count?

**Review by nonword — APPROVED** [review]
Immense improvements

File: standards/technical-approach.md
Comment (samanthaandrews) [reply standards/technical-approach.md]: Good point. I added this clarification. Agree that addendums are helpful and they shouldn't count toward the 10-12 page limit.

File: standards/technical-approach.md
Comment (samanthaandrews) [reply standards/technical-approach.md]: Yeah I agree. Updated.

**Pull request merged**
